crates/server/src/server/client.rs

```diff
@@ -0,0 +1,169 @@
+use std::any::TypeId;
+
+use lsp_server::{Notification, RequestId};
+use rustc_hash::FxHashMap;
+use serde_json::Value;
+
+use super::{schedule::Task, ClientSender};
+
+type ResponseBuilder<'s> = Box<dyn FnOnce(lsp_server::Response) -> Task<'s>>;
+
+pub(crate) struct Client<'s> {
+    notifier: Notifier,
+    responder: Responder,
+    pub(super) requester: Requester<'s>,
+}
+
+#[derive(Clone)]
+pub(crate) struct Notifier(ClientSender);
+
+#[derive(Clone)]
+pub(crate) struct Responder(ClientSender);
+
+pub(crate) struct Requester<'s> {
+    sender: ClientSender,
+    next_request_id: i32,
+    response_handlers: FxHashMap<lsp_server::RequestId, ResponseBuilder<'s>>,
+}
+
+impl Client<'_> {
+    pub(super) fn new(sender: ClientSender) -> Self {
+        Self {
+            notifier: Notifier(sender.clone()),
+            responder: Responder(sender.clone()),
+            requester: Requester {
+                sender,
+                next_request_id: 1,
+                response_handlers: FxHashMap::default(),
+            },
+        }
+    }
+
+    pub(super) fn notifier(&self) -> Notifier {
+        self.notifier.clone()
+    }
+
+    pub(super) fn responder(&self) -> Responder {
+        self.responder.clone()
+    }
+}
+
+#[allow(dead_code)] // we'll need to use `Notifier` in the future to send notifs to the client
+impl Notifier {
+    pub(crate) fn notify<N>(&self, params: N::Params) -> anyhow::Result<()>
+    where
+        N: lsp_types::notification::Notification,
+    {
+        let method = N::METHOD.to_string();
+
+        let message = lsp_server::Message::Notification(Notification::new(method, params));
+
+        self.0.send(message)
+    }
+
+    pub(crate) fn notify_method(&self, method: String) -> anyhow::Result<()> {
+        self.0
+            .send(lsp_server::Message::Notification(Notification::new(
+                method,
+                Value::Null,
+            )))
+    }
+}
+
+impl Responder {
+    pub(crate) fn respond<R>(
+        &self,
+        id: RequestId,
+        result: crate::server::Result<R>,
+    ) -> anyhow::Result<()>
+    where
+        R: serde::Serialize,
+    {
+        self.0.send(
+            match result {
+                Ok(res) => lsp_server::Response::new_ok(id, res),
+                Err(crate::server::api::Error { code, error }) => {
+                    lsp_server::Response::new_err(id, code as i32, format!("{error}"))
+                }
+            }
+            .into(),
+        )
+    }
+}
+
+impl<'s> Requester<'s> {
+    /// Sends a request of kind `R` to the client, with associated parameters.
+    /// The task provided by `response_handler` will be dispatched as soon as the response
+    /// comes back from the client.
+    pub(crate) fn request<R>(
```

Comment:
Typically the server _receives_ requests/notifs from the client, but in a few places we will also _send_ requests/notifs to the client, possibly in the middle of some other request.

That is handled too! We have access to a `client.requester` that we can send along a request through. When the response comes back, a handler that we provided is immediately run. The response is matched up to our handler through a `RequestId` that is created during the request.

We use this to register dynamic capabilities, like `didChangeWatchedFiles` server support